### PR TITLE
Fix stuff related to the causal_mask in CodeGen.

### PR DESCRIPTION
**Pull request by GeneZC**
1. Line 613, `_keys_to_ignore_on_load_missing  =  [r"h\.\d+\.attn\.masked_bias", r"h\.\d+\.attn\.bias"]` => `_keys_to_ignore_on_load_missing  =  [r"h\.\d+\.attn\.causal_mask"]` to load correctly from CodeGen checkpoint without `causal_mask`.
2. Line 152, `causal_mask = self.causal_mask[:, :, key_length - query_length : key_length, :key_length]
` => `causal_mask = self.causal_mask[:, :, key_length - query_length : key_length, :key_length].bool()
` to alleviate potential user warning saying like `UserWarning: where received a uint8 condition tensor. This behavior is deprecated and will be removed in a future version of PyTorch. Use a boolean condition instead.`.

# What does this PR do?

Fixes # (issue)

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by sgugger** [comment]
cc @ArthurZucker and @younesbelkada

**Review by younesbelkada — COMMENTED** [review]
Thanks a lot for your PR! 

At a first glance, the line:
```
causal_mask = self.causal_mask[:, :, key_length - query_length : key_length, :key_length].bool()
```
Seems to be not needed since https://github.com/huggingface/transformers/pull/21384 should take care of that (cc @ArthurZucker 👀 ).
But after thinking a bit, since the buffer that are on the Hub are already stored as `uint8` tensors, this might be needed. I need to dig more, (meanwhile @sgugger & arthur, if you have a better idea let me know)

However I did not really followed your second change, could you elaborate more on why you removed `masked.bias` & `attn\.bias`?

Thanks!

**Comment by younesbelkada** [comment]
Following up on my comment, here is a script I ran: 

```python

import torch
import torch.nn as nn

class DummyModelUint(nn.Module):
    def __init__(self):
        super().__init__()
        self.linear = nn.Linear(10, 10)
        self.register_buffer("attention_mask", torch.ones(10, 10, dtype=torch.uint8))

    def forward(self, x):
        return self.linear(x)

class DummyModelBool(nn.Module):
    def __init__(self):
        super().__init__()
        self.linear = nn.Linear(10, 10)
        self.register_buffer("attention_mask", torch.ones(10, 10, dtype=torch.bool))

    def forward(self, x):
        return self.linear(x)

model_uint8 = DummyModelUint()
state_dict = model_uint8.state_dict()

model_bool = DummyModelBool()
model_bool.load_state_dict(state_dict)

print(model_bool.attention_mask.dtype)
>>> torch.bool
```
So even though you save `uint8` buffers on the state dict, if you load it through a model that has `torch.bool` buffer it will overwrite it in `torch.bool`. So I don't think the first fix is needed if we merge #21384 ! 
However your second fix might make sense, just need more clarification ;)

**Comment by GeneZC** [comment]
@younesbelkada 
Thanks for your review.

1. Removing `[r"h\.\d+\.attn\.masked_bias", r"h\.\d+\.attn\.bias"]` basically due to there are not `masked_bias` or `bias` in `attn`. I conjecture the reason why they leave these two terms here is that they copy part of the code from GPTJ (https://github.com/huggingface/transformers/blob/main/src/transformers/models/gptj/modeling_gptj.py#L723).
2. Introducing `.bool()` due to it will raise a warning when later execute the `torch.where` (https://github.com/huggingface/transformers/blob/main/src/transformers/models/codegen/modeling_codegen.py#L165).

**Comment by ArthurZucker** [comment]
But the causal mask should not be a `bool`. The only bool is going the be the attention mask that we feed the network, which is then used to created the causal mask by multiplying with `torch.finfo(...).min`

**Review by younesbelkada — COMMENTED** [review]
Thank you @GeneZC ! The explanation of 1. makes sense to me; `bias` refer to this:https://github.com/huggingface/transformers/blob/21a2d900eceeded7be9edc445b56877b95eda4ca/src/transformers/models/gptj/modeling_gptj.py#L91 which is present in GPTJ but not in CodeGen. Same for `masked_bias`. I think let's address 2. in #21384 , ok for me to merge if 2. is removed!

**Review by younesbelkada — APPROVED** [review]
Thanks a lot for taking care of this!

**Review by sgugger — APPROVED** [review]
Thanks for your contribution!
